### PR TITLE
Add actions CRUD [CLI-185]

### PR DESCRIPTION
**Pull request by Widcket**
### Description

This PR adds the following commands:
- `auth0 actions list`
- `auth0 actions show`
- `auth0 actions create`
- `auth0 actions update`
- `auth0 actions delete`
- `auth0 actions open`

Screen Shot 2021-06-21 at 12 51 40

#### Create

https://user-images.githubusercontent.com/5055789/122792625-b93f7780-d290-11eb-8dc1-c9ad52907914.mov

#### Update

https://user-images.githubusercontent.com/5055789/122796702-12a9a580-d295-11eb-9ec2-87b998397335.mov

### Testing

- [ ] This change adds test coverage for new/changed/fixed functionality

### Checklist

- [ ] I have added documentation for new/changed functionality in this PR or in auth0.com/docs
- [x] All active GitHub checks for tests, formatting, and security are passing
- [ ] The correct base branch is being used, if not `master`

### CODE REVIEW

File: internal/cli/actions.go [hunk 1]
```diff
@@ -0,0 +1,540 @@
...
+
+			if inputs.Name == "" {
+				inputs.Name = current.GetName()
+			}
+
+			if inputs.Trigger == "" && len(current.SupportedTriggers) > 0 {
+				inputs.Trigger = current.SupportedTriggers[0].GetID()
+			}
+
+			unprexifedVersion := strings.TrimPrefix(version, "v")
```
Comment (Widcket) [inline internal/cli/actions.go:312]: Actions update fails with the prefixed version that comes from the triggers endpoint (eg `v2`):
Screen Shot 2021-06-21 at 12 25 38
Screen Shot 2021-06-21 at 12 25 48
That is not the case for actions create, so it's probably a bug..

File: commander.yaml [hunk 2]
```diff
@@ -276,7 +288,7 @@ tests:
         - IDENTIFIER            http://integration-test-api-def2
         - SCOPES                read:todos write:todos
         - TOKEN LIFETIME        86400
-        - ALLOW OFFLINE ACCESS  false
+        - ALLOW OFFLINE ACCESS  ✗
```
Comment (jfatta) [inline commander.yaml:291]: cool. are we ok with the --format=JSON output? true || false ?

File: internal/auth/mock/auth.go [hunk 3]
```diff
@@ -5,34 +5,49 @@
 package mock
 
 import (
-	gomock "github.com/golang/mock/gomock"
 	reflect "reflect"
+
+	gomock "github.com/golang/mock/gomock"
 )
 
-// MockSecretStore is a mock of SecretStore interface
```
Comment (jfatta) [inline internal/auth/mock/auth.go:12]: nit: isn't this file auto generated? why the tiny changes? If it was part of a "format-on-save" action, I'd suggest reverting it because it will cause unnecessary changes once we generate mocks again

File: internal/auth0/log_mock.go [hunk 4]
```diff
@@ -5,74 +5,75 @@
...
 
 import (
+	reflect "reflect"
+
 	gomock "github.com/golang/mock/gomock"
 	management "gopkg.in/auth0.v5/management"
-	reflect "reflect"
 )
 
-// MockLogAPI is a mock of LogAPI interface
```
Comment (jfatta) [inline internal/auth0/log_mock.go:13]: there are a lot of this mock file changes, I wonder if you used a new version. Do we have the tool vendored?

File: internal/cli/actions.go [hunk 5]
```diff
@@ -0,0 +1,540 @@
...
+		Name:       "Trigger",
+		LongForm:   "trigger",
+		ShortForm:  "t",
+		Help:       "Trigger of the action. At this time, an action can only target a single trigger at a time.",
+		IsRequired: true,
+	}
+
+	actionCode = Flag{
+		Name:       "Code",
+		LongForm:   "scode",
```
Comment (jfatta) [inline internal/cli/actions.go:39]:
```suggestion
		LongForm:   "code",
```

File: internal/auth0/connection.go [hunk 6]
```diff
@@ -1,4 +1,4 @@
-//go:generate mockgen -source=connection.go -destination=connection.go -package=auth0
+//go:generate mockgen -source=connection.go -destination=connection_mock.go -package=auth0
```
Comment (jfatta) [inline internal/auth0/connection.go:1]: great catch

File: internal/auth0/anomaly.go [hunk 7]
```diff
@@ -1,3 +1,5 @@
+//go:generate mockgen -source=anomaly.go -destination=anomaly_mock.go -package=auth0
```
Comment (jfatta) [inline internal/auth0/anomaly.go:1]: are we using this new mock already? If not, I prefer to avoid its generation

File: internal/auth0/branding_mock.go [hunk 8]
```diff
@@ -0,0 +1,129 @@
+// Code generated by MockGen. DO NOT EDIT.
```
Comment (jfatta) [inline internal/auth0/branding_mock.go:1]: do we need this mock?

File: internal/cli/data/action-template-credentials-exchange.js [hunk 9]
```diff
@@ -0,0 +1,8 @@
+/**
```
Comment (jfatta) [inline internal/cli/data/action-template-credentials-exchange.js:1]: I'm not familiar with the /data/ files usage. Do you think we could have /cli/data/README.md explaining briefly?

File: internal/cli/root.go [hunk 10]
```diff
@@ -119,6 +119,7 @@ func Execute() {
 	rootCmd.AddCommand(brandingCmd(cli))
 	rootCmd.AddCommand(rolesCmd(cli))
 	rootCmd.AddCommand(ipsCmd(cli))
+	rootCmd.AddCommand(actionsCmd(cli))
```
Comment (jfatta) [inline internal/cli/root.go:122]: nit: I'll put this close to 	`rootCmd.AddCommand(rulesCmd(cli))`

File: internal/display/apis.go [hunk 11]
```diff
@@ -112,12 +112,12 @@ func (r *Renderer) ApiUpdate(api *management.ResourceServer) {
 func makeApiView(api *management.ResourceServer) (*apiView, bool) {
 	scopes, scopesTruncated := getScopes(api.Scopes)
 	view := &apiView{
-		ID:            auth0.StringValue(api.ID),
-		Name:          auth0.StringValue(api.Name),
-		Identifier:    auth0.StringValue(api.Identifier),
-		Scopes:        auth0.StringValue(scopes),
-		TokenLifetime: auth0.IntValue(api.TokenLifetime),
-		OfflineAccess: auth0.BoolValue(api.AllowOfflineAccess),
+		ID:            ansi.Faint(api.GetID()),
```
Comment (jfatta) [inline internal/display/apis.go:115]: just a personal preference: I'd like to avoid multiple command changes that are unrelated. As trivial as those can be, smaller PRs are easier to review, revert (if needed 🤞🏼 ), trace old changes if needed (troubleshooting).

**Review by jfatta — COMMENTED** [review]
looking good 🙇🏼

File: commander.yaml [hunk 12]
```diff
@@ -20,6 +23,15 @@ tests:
   auth0 rules list:
     exit-code: 0
 
+  auth0 actions list:
```
[the remark this answers is not in the document]
Comment (Widcket) [reply commander.yaml:26]: It's a good idea, but will make it costlier to contribute, and we don't get many contributions now.

File: commander.yaml (hunk 2 above)
Comment (Widcket) [reply commander.yaml:291]: Yes, the JSON output retains the `true` and `false`.

File: internal/cli/actions.go (hunk 5 above)
Comment (Widcket) [reply internal/cli/actions.go:39]: Good catch, thanks!

File: internal/auth0/anomaly.go (hunk 7 above)
Comment (Widcket) [reply internal/auth0/anomaly.go:1]: Ok, I'll remove the mocks.

File: internal/cli/root.go (hunk 10 above)
Comment (Widcket) [reply internal/cli/root.go:122]: You're right, that determines the order commands will show up on the help.

File: internal/display/apis.go (hunk 11 above)
Comment (Widcket) [reply internal/display/apis.go:115]: Yes, that's true. However, it's far easier to get a reviewer for a single PR. However, I'll try to split these into their own PRs in the future.

File: internal/cli/data/action-template-credentials-exchange.js (hunk 9 above)
Comment (Widcket) [reply internal/cli/data/action-template-credentials-exchange.js:1]: Sure, added in https://github.com/auth0/auth0-cli/pull/316/commits/4023c615561e84a418e494f6c4b9b2d4737e6a6b

File: internal/auth0/log_mock.go (hunk 4 above)
Comment (Widcket) [reply internal/auth0/log_mock.go]: Yes, I used the latest version of mockgen. It's not vendored:
Screen Shot 2021-06-21 at 20 45 38
I removed the newly generated mocks and reverted the exiting ones.
Good catch!

File: internal/auth0/anomaly.go (hunk 7 above)
Comment (Widcket) [reply internal/auth0/anomaly.go:1]: Done. The new ones were removed in https://github.com/auth0/auth0-cli/pull/316/commits/46cb25c7903ba70db43d8a58f7fb24bdebb45a89, and the modified ones (because of different version of mockgen) were reverted in https://github.com/auth0/auth0-cli/pull/316/commits/e1cf52dfea62b9fc7a456293b03b466f2842d95e and https://github.com/auth0/auth0-cli/pull/316/commits/675930fad27b53ea634b529b4aed0252ee929817.

File: internal/auth0/branding_mock.go (hunk 8 above)
Comment (Widcket) [reply internal/auth0/branding_mock.go]: Removed in https://github.com/auth0/auth0-cli/pull/316/commits/46cb25c7903ba70db43d8a58f7fb24bdebb45a89.

File: internal/cli/root.go (hunk 10 above)
Comment (Widcket) [reply internal/cli/root.go]: Changed in https://github.com/auth0/auth0-cli/pull/316/commits/2e653f5e2bc005ce8cd614c8243f4bd3199b992f.

File: internal/auth/mock/auth.go (hunk 3 above)
Comment (Widcket) [reply internal/auth/mock/auth.go]: Reverted in https://github.com/auth0/auth0-cli/pull/316/commits/6d0b3efcfe9fe399a43b93dc202291e3ed6496e6.

File: internal/cli/actions.go (hunk 1 above)
Comment (Widcket) [reply internal/cli/actions.go]: Removed in https://github.com/auth0/auth0-cli/pull/316/commits/ae517fa60da115323ded70d0c35765b81a57348f, as it does not fail anymore.

File: commander.yaml (hunk 12 above)
[the remark this answers is not in the document]
Comment (jfatta) [reply commander.yaml:26]: good point

File: internal/auth/mock/auth.go (hunk 3 above)
Comment (jfatta) [reply internal/auth/mock/auth.go]: thanks 🙇🏼

File: internal/auth0/anomaly.go (hunk 7 above)
Comment (jfatta) [reply internal/auth0/anomaly.go:1]: great work, thanks

File: internal/display/apis.go (hunk 11 above)
Comment (jfatta) [reply internal/display/apis.go:115]: thanks 🙇🏼

Comment (jfatta) [reply internal/display/apis.go:115]: I believe smaller PRs would also lower the barrier for new reviewers, give feedback on a ~50 files PR can be overwhelming 😅

File: internal/cli/root.go [hunk 13]
```diff
@@ -106,19 +106,20 @@ func Execute() {
...
 	rootCmd.AddCommand(tenantsCmd(cli))
-	rootCmd.AddCommand(usersCmd(cli))
 	rootCmd.AddCommand(appsCmd(cli))
+	rootCmd.AddCommand(usersCmd(cli))
 	rootCmd.AddCommand(rulesCmd(cli))
+	rootCmd.AddCommand(actionsCmd(cli))
 	rootCmd.AddCommand(apisCmd(cli))
+	rootCmd.AddCommand(rolesCmd(cli))
+	rootCmd.AddCommand(brandingCmd(cli))
+	rootCmd.AddCommand(ipsCmd(cli))
```
Comment (jfatta) [inline internal/cli/root.go:119]: nice

**Review by jfatta — APPROVED** [review]
awesome work, thanks
